logdna/request.go

```diff
@@ -0,0 +1,139 @@
+package logdna
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"time"
+)
+
+type httpRequest func(string, string, io.Reader) (*http.Request, error)
+type bodyReader func(io.Reader) ([]byte, error)
+type jsonMarshal func(interface{}) ([]byte, error)
+type httpClientInterface interface {
+	Do(*http.Request) (*http.Response, error)
+}
+
+// Client used to make HTTP requests to the configuration api
+type requestConfig struct {
+	serviceKey  string
+	httpClient  httpClientInterface
+	apiURL      string
```

Comment:
I think if this contained the `baseURL` (`Host`, aka `URL` from the previous code), you could construct the full `apiURL` inside of e.g `CreateAlert`. then you wouldn't need to pass in `config.Host` for every CRUD API method

Reply:
Right, so that's the intention.  I'm not sure if you *meant* `CreateAlert`, but all of those "createXXX" functions are being removed.  We will always have to pass `baseURL` in some form, but it doesn't need to be a property of the `requestConfig`.  This is the flow now:

1) baseURL is part of "provider config" and is static.
2) `pc` is passed by ref to `newRequestConfig` which abstracts the struct creation so we can mock
3) `newRequestConfig` creates the full `apiURL` using the `pc.baseURL` and the provided `uri`

```go
	req := newRequestConfig(
		pc,
		"POST",
		"/v1/config/view",
		view,
	)
```